File: src/TracePoints.js
var T     = require ( "./Tango" ) ;
var Event = require ( "./Event" ) ;
var Log   = require ( "./LogFile" ) ;

/**
 * { function_description }
 *
 * @class      TracePoint
 * @param      {string}  name    name of the tracepoint
 * @param      {boolean}  active  true|false
 */
var TracePoint = function ( name, active )
{
  this.active        = !! active ;
  this.name          = name ;
  this.mode          = "" ;
  this.store         = null ;
  this.title         = "" ;
  this.tracer        = null ;
  this.includeSystem = false ;
};
/**
 * { function_description }
 *
 * @method     setTitle
 * @param      {<type>}  title   { description }
 */
TracePoint.prototype.setTitle = function ( title )
{
  this.title = title ;
  return this ;
};
/**
 * { function_description }
 *
 * @method     log
 * @param      {String}  value   { description }
 */
TracePoint.prototype.log = function ( value )
{
  if ( ! this.active )
  {
    return false ;
  }
  var tracer = this.tracer ;
  if ( ! tracer )
  {
    tracer = this.store.tracer ;
  }
  s = "" ;
  if ( value instanceof Event )
  {
    if ( value.getName() === "system" && ! this.includeSystem )
    {
      return false ;
    }
    if ( this.title )
    {
      s += this.title ;
      s += "\n" ;
    }
    var mode = this.mode ;
    if ( ! mode ) mode = 'shb' ; // status header and body
    if ( mode === 'a' )
    {
      s += T.toString ( value ) ;
      s += "\n" ;
    }
    if ( mode.indexOf ( 'h' ) >= 0 )
    {
      s += value.getName() + "/" + value.getType() ;
      s += "\n" ;
    }
    if ( mode.indexOf ( 'u' ) >= 0 )
    {
      s += T.toString ( value.user ) ;
      s += "\n" ;
    }
    if ( mode.indexOf ( 'c' ) >= 0 )
    {
      s += T.toString ( value.control ) ;
      s += "\n" ;
    }
    if ( mode.indexOf ( 's' ) >= 0 )
    {
      if ( value.getStatus() )
      {
        s += T.toString ( value.getStatus() ) ;
        s += "\n" ;
      }
    }
    if ( mode.indexOf ( 'b' ) >= 0 )
    {
      s += T.toString ( value.body ) ;
      s += "\n" ;
    }
  }
  else
  {
    if ( this.title )
    {
      s += this.title ;
      s += "\n" ;
    }
    s += T.toString ( value ) ;
  }
  tracer ( s ) ;
  return true ;
};
TracePoint.prototype.isActive = function()
{
  return this.active ;
};
TracePoint.prototype.action = function ( action )
{
  if ( ! action )
  {
    return ;
  }
  if ( typeof action.system !== 'undefined' )
  {
    if ( action.system == "true" )
    {
      this.includeSystem = true ;
    }
    else
    {
      this.includeSystem = false ;
    }
  }
}
/**
 * { function_description }
 *
 * @class      TracePointStore
 * @param      {<type>}  name    { description }
 */
var TracePointStore = function ( name )
{
  this.points        = {} ;
  this.name          = name ? name : "" ;
  this.localTracer   = Log.logln.bind ( Log ) ;
  this.remoteTracer  = null ;
  this.tracer        = this.localTracer ;
};
TracePointStore.prototype.getName = function()
{
  return this.name ;
};
TracePointStore.prototype.log = function ( storeName, value )
{
  var tp = this.points[storeName] ;
  if ( ! tp )
  {
    return ;
  }
  return tp.log ( value ) ;
};
TracePointStore.prototype.add = function ( tp, isActive )
{
  if ( tp instanceof TracePoint )
  {
    this.points[tp.name] = tp ;
    tp.active = T.getProperty ( tp.name, tp.active ) ;
  }
  else
  {
    var name  = tp ;
    tp        = new TracePoint ( name, !!isActive ) ;
    tp.active = T.getProperty ( name, tp.active ) ;
    this.points[name] = tp ;
  }
  tp.store = this ;
  return tp ;
};
TracePointStore.prototype.getTracePoint = function ( name )
{
  return this.points[name] ;
};
TracePointStore.prototype.remove = function ( name )
{
  if ( this.points[name] )
  {
    this.points[name].store = null ; 
  }
  delete this.points[name] ;
};
TracePointStore.prototype.action = function ( action )
{
  var tp = null ;
  var i, j, k ;
  if ( action && action.output )
  {
    if ( action.output === 'remote' )
    {
      this.tracer = this.remoteTracer ;
    }
    else
    {
      this.tracer = this.localTracer ;
    }
  }
  if ( action && action.points )
  {
    for ( i = 0 ; i < action.points.length ; i++ )
    {
      var item = action.points[i] ;
      if ( item.name === '*' )
      {
        for ( k in this.points )
        {
          if ( item.state === 'on' ) this.points[k].active = true ;
          if ( item.state === 'off' ) this.points[k].active = false ;
          if ( item.state === 'toggle' ) this.points[k].active = ! this.points[k].active ;
          this.points[k].mode = item.mode ;
          this.points[k].action ( item )
        }
        continue ;
      }
      tp = this.points[item.name] ;
      if ( tp )
      {
        if ( item.state === 'on' ) tp.active = true ;
        if ( item.state === 'off' ) tp.active = false ;
        if ( item.state === 'toggle' ) tp.active = ! tp.active ;
        tp.mode = item.mode ;
        tp.action ( item )
      }
    }
  }
  var result ;
  {
    result = { name: this.getName(), list: [] } ;
    result["output"] = this.tracer == this.localTracer ? "local" : "remote" ;
    for ( k in this.points )
    {
      result.list.push ( { name:this.points[k].name
                         , active:this.points[k].active
                         , system:this.points[k].includeSystem
                         }
                       );
    }
  }
  return result ;
};
if ( typeof org === 'undefined' ) org = {} ;
if ( typeof org.gessinger === 'undefined' ) org.gessinger = {} ;
if ( typeof org.gessinger.tangojs === 'undefined' ) org.gessinger.tangojs = {} ;

if ( ! org.gessinger.tangojs.TracePoints )
{
  org.gessinger.tangojs.TracePoints =
  { list: []
  , store: {}
  , getStore: function getStore ( name )
  {
    if ( ! name ) name = "" ;
    if ( ! this.store[name] )
    {
      this.store[name] = new TracePointStore ( name ) ;
    }
    return this.store[name] ;
  }
  , TracePoint: TracePoint
  } ;
}
module.exports = org.gessinger.tangojs.TracePoints ;

File: python/TracePoints.py
#!/usr/bin/env python

try:
  from cStringIO import StringIO
except ImportError:
  from io import StringIO

from io import BytesIO

import json
import inspect
import time
import datetime
import socket
import sys
import os
import threading
import types
import numbers
import collections
import resource

from gepard import Event, util

def localTracer(t):
  print ( t )

class TracePoint ( object ):
  def __init__ ( self, name="", active=False ):
    self.active = active
    self.name   = name
    self.mode   = "hb"
    self.store  = None
    self.title  = None
    self.tracer = None
  
  def setTitle ( self, title ):
    self.title = title

  def log ( self, value ):
    if not self.active:
      return
    tracer = self.tracer
    if tracer == None:
      tracer = self.store.tracer
    if isinstance ( value, Event ):
      e = value
      if self.title != None:
        tracer ( self.title )
      mode = self.mode
      if mode == None:
        mode = 'hb'
      if mode == 'a':
        tracer ( e )
      if mode.find ( 'h' ) >= 0: tracer ( e.getName() + "/" + e.getType() )
      if mode.find ( 'u' ) >= 0: tracer ( e.user )
      if mode.find ( 'c' ) >= 0: tracer ( e.control )
      if mode.find ( 'b' ) >= 0: tracer ( e.body )
    else:
      t = str ( value ) ;
      if self.title != None:
        tracer ( self.title )
      tracer ( t )
  def isActive ( self ):
    return self.active

class TracePointStore ( object ):
  def __init__ ( self, name="" ):
    self.points       = {}
    self.localTracer  = localTracer
    self.remoteTracer = None
    self.tracer       = self.localTracer
    self.name         = name
  def getName ( self ):
    return self.name

  def add ( self, tp, isActive=False ):
    if isinstance ( tp, TracePoint ):
      self.points[tp.name] = tp
      tp.active = util.getProperty ( tp.name, tp.active )
    else:
      name              = tp
      tp                = TracePoint ( name, isActive )
      tp.active         = util.getProperty ( name, tp.active )
      self.points[name] = tp
      tp.store          = self
      return tp

  def getTracePoint ( self, name ):
    return self.points.get ( name )

  def remove( self, name ):
    if not isinstance ( name, basestring):
      return
    if name in self.points:
      del self.points[name]

  def action ( self, action=None ):
    if isinstance ( action, dict ):
      if "output" in action:
        output = action["output"]
        if output == "remote":
          self.tracer = self.remoteTracer
        else:
          self.tracer = self.localTracer

      if "points" in action:
        points = action["points"]
        for i in range ( 0, len ( points ) ):
          item = points[i]
          if not "name" in item:
            continue
          mode = item.get ( "mode" )
          name = item.get ( "name" )
          if name == "*":
            state = item.get ( "state" )
            if state == None:
              continue
            for k in self.points:
              point = self.points[k]
              if state == 'on': point.active = True
              if state == 'off': point.active = False
              if state == 'toggle': point.active = not point.active
              if mode != None:
                point.mode = mode
            continue

          tp = self.points[name] ;
          if tp == None:
            continue
          state = item.get ( "state" )
          if state == None:
            continue
          if state == 'on': tp.active = True
          if state == 'off': tp.active = False
          if state == 'toggle': tp.active = not tp.active
          if mode != None:
            point.mode = mode
    result = {}
    result["name"] = self.getName()
    list = []
    result["list"] = list
    for k in self.points:
      point = self.points[k]
      list.append ( { "name":point.name, "active":point.active } )
    return result
